docs/lang/articles/debug/debugging.md

```diff
@@ -146,40 +104,90 @@ def inside_taichi_scope():
     for i in range(4):
         ti.static_print(i.dtype)
         # => DataType.int32
-        # will only print once
+        # Only print once
 ```
 
-Unlike `print`, `ti.static_print` will only print the expression once at
-compile-time, and therefore it has no runtime cost.
+`ti.static_print` behaves similarly with `print` in the Python scope. The difference is that unlike `print`, `ti.static_print` only prints the expression once at compile time, thus incurring no runtime cost.
 
 ## Serial execution
 
-The automatic parallelization feature of Taichi may lead to
-non-deterministic behaviors. For debugging purposes, it may be useful to
-serialize program execution to get repeatable results and to diagnose
-data races. When running your Taichi program on CPUs, you can initialize
-Taichi to use a single thread with `cpu_max_num_threads=1`, so that the
-whole program becomes serial and deterministic. For example,
+Taichi's automatic parallelization mechanism may lead to non-deterministic behaviors. For debugging purposes, serializing program execution may be useful to get repeatable results and diagnose data races. You can serialize either an entire Taichi program or a specific for loop.
 
-```
+### Serialize an entire Taichi program
+
+If you choose CPU as the backend, you can serialize the program with `cpu_max_num_threads=1` when initiating Taichi, so that the whole program runs on a single thread and becomes deterministic. For example:
+
+```python
 ti.init(arch=ti.cpu, cpu_max_num_threads=1)
 ```
 
-If your program works well in serial but not in parallel, check
-parallelization-related issues such as data races.
+If your program works well in serial but fails in parallel, check parallelization-related issues, such as data races.
+
+### Serialize a specified parallel for loop
 
-## Runtime `assert` in Taichi-scope
+By default, Taichi automatically parallelizes the for loop at the outermost scope. But some scenarios require serial exeution. In this case, you can prevent automatic parallelization with `ti.loop_config(serialize=True)`:
```

Comment:
Taichi automatically parallelizes the for loops at the outermost scope in a Taichi kernel. But xxx